File: apps/ios/Exponential/UI/Actions/CreateActionSheet.swift
import ExpCore
import ExpUI
import GRDB
import SwiftUI

// EXP-615: "New action" is its own sheet. It used to be a MODE of the
// Start-coding sheet (EXP-431's `createActionMode`), which meant the creation
// flow inherited a subject switch it had to hide, an action picker it had to
// suppress, and — once suggestions could seed an automation (EXP-583) — a
// trigger block that only appeared for one entry point. Now creation looks
// like every other create form on mobile: icon + name, description,
// repository, an ALWAYS-visible automation row that pushes a detail form, and
// the shared launch options for the run that authors it.
//
// Creation itself is unchanged: the builtin "Create action" run does the
// writing (there is no manual action form on any client since EXP-257), so
// submitting starts that action on a machine with the typed inputs, and a
// configured automation rides the description as the byte-identical
// `AutomationNote` block the creator agent copies into
// `exponential_automations_create`.
struct CreateActionSheet: View {
    let teamId: String
    /// Machines a builtin action run can be sent to (online with a runnable
    /// agent — EXP-672 dropped the action capability filters).
    let devices: [SteerDevice]
    /// Automation-capable machines (cap `automations`), OFFLINE INCLUDED —
    /// the automation binds independently of whatever machine runs the
    /// creator session, and a sleeping box still owns the binding.
    let automationDevices: [SteerDevice]
    /// EXP-530 "Use": the seeded description + icon.
    var prefillDescription: String = ""
    var prefillIcon: String = ""
    /// EXP-583: an "Action + automation" seed's trigger. Unlike before, an
    /// automation can also be configured from scratch — this only PRE-fills.
    var prefillAutomation: AutomationTrigger?
    /// The builtin create run: device, action, options, resolved inputs.
    let onSubmit: (SteerDevice, ActionDto, SteerStartOptions, [String: String]) -> Void

    @Environment(AppDependencies.self) private var deps
    @Environment(\.accountId) private var accountId
    @Environment(\.dismiss) private var dismiss

    @State private var name = ""
    @State private var descriptionText = ""
    @State private var icon = ""
    @State private var repoId = ""
    @State private var repos: [TeamRepo] = []

    // Automation (optional): `hasAutomation` is the row's on/off state, the
    // draft its when-part, and the device/agent pins its how-part.
    @State private var hasAutomation = false
    @State private var draft = AutomationDraft()
    @State private var filterOptions = AutomationFilterOptions()
    @State private var automationDeviceId = ""
    @State private var automationAgent = ""
    @State private var automationModel = LaunchVocabulary.cliDefault
    @State private var automationEffort = LaunchVocabulary.cliDefault

    // The creator RUN's own options, seeded from the resolved machine's
    // advertised defaults (EXP-437) exactly like the Start-coding sheet.
    @State private var deviceId: String?
    @State private var agent = "claude"
    @State private var model = ""
    @State private var effort = LaunchVocabulary.cliDefault
    @State private var ultracode = false
    @State private var planMode = false
    @State private var seeded = false
    @State private var lastSeededDeviceId: String?

    /// The action being run: the "Create action" builtin, whose input defs
    /// are the contract for what this form collects.
    private var builtin: ActionDto {
        ActionDto.builtinCreateAction(teamId: teamId)
    }

    private var device: SteerDevice? {
        if let deviceId, let match = devices.first(where: { $0.deviceId == deviceId }) {
            return match
        }
        // EXP-622: the caller's default machine, when it is still a candidate.
        return devices.first(where: \.isDefaultDevice) ?? devices.first
    }

    private var availableAgents: [String] {
        let supported = device?.agentIds ?? ["claude"]
        let ordered = DomainContract.codingAgentValues.filter { supported.contains($0) }
        return ordered.isEmpty ? ["claude"] : ordered
    }

    private var canSubmit: Bool {
        device != nil
            && !descriptionText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            && descriptionText.count <= DomainContract.actionInputTextMax
            && name.count <= DomainContract.actionInputTextMax
    }

    var body: some View {
        // The inner NavigationStack stays for the automation push; its own bar
        // is hidden on the root so the sheet header is the only one (EXP-687).
        GlassSheetChrome(
            title: "New action",
            height: .full,
            content: {
                NavigationStack {
                    Form {
                        identitySection
                        descriptionSection
                        repositorySection
                        automationSection
                        LaunchOptionsSection(
                            variant: .launch,
                            devices: devices,
                            deviceId: deviceBinding,
                            noDeviceNote: noDeviceNote,
                            availableAgents: availableAgents,
                            agent: agent,
                            onAgentChange: selectAgent,
                            model: $model,
                            effort: $effort,
                            ultracode: $ultracode,
                            planMode: $planMode
                        )
                    }
                    // EXP-603: the sheet's own background shows through the
                    // grouped list; rows carry the glass fill.
                    .scrollContentBackground(.hidden)
                    .listSectionSpacing(8)
                    // EXP-594: white control tint — system blue is retired.
                    .tint(DesignTokens.Palette.primary)
                    .toolbar(.hidden, for: .navigationBar)
                }
            },
            primaryAction: {
                GlassSubmitButton("Create", enabled: canSubmit) {
                    submit()
                }
            }
        )
        .accessibilityElement(children: .contain)
        .accessibilityIdentifier("create-action-sheet")
        .onAppear { seed() }
        .task { await load() }
    }

    // MARK: - Form sections

    /// Icon + name on ONE row (web/desktop parity): the glyph the action will
    /// wear, and an optional name — blank lets the agent pick one. EXP-694
    /// (S7): identical to the edit sheet's identity row, headerless — the
    /// placeholder IS the title on every editor surface now.
    private var identitySection: some View {
        Section {
            HStack(spacing: 12) {
                IconPicker(selection: $icon, allowsNone: true)
                TextField("Name (optional)", text: $name)
                    .accessibilityIdentifier("create-action-name")
            }
        }
        .listRowBackground(glassFormRowFill)
    }

    private var descriptionSection: some View {
        Section {
            TextField("What should this action do?", text: $descriptionText, axis: .vertical)
                .lineLimit(4...10)
                .accessibilityIdentifier("create-action-description")
        }
        .listRowBackground(glassFormRowFill)
    }

    @ViewBuilder
    private var repositorySection: some View {
        Section {
            GlassPickerRow(
                "Repository",
                selection: $repoId,
                options: [""] + repos.map(\.id),
                label: { id in
                    guard !id.isEmpty else { return "None" }
                    return repos.first { $0.id == id }?.fullName ?? id
                }
            )
        }
        .listRowBackground(glassFormRowFill)
    }

    /// The always-visible automation row — an issue-row shaped summary that
    /// pushes the detail form. "No automation" is the empty state; a
    /// configured one reads as its trigger sentence plus the bound machine.
    private var automationSection: some View {
        Section {
            NavigationLink {
                automationDetail
            } label: {
                HStack(spacing: 12) {
                    AppIcon(AppIcons.actionAutomation, size: AppIcon.Size.medium)
                        .foregroundStyle(.white.opacity(TextOpacity.secondary))
                    VStack(alignment: .leading, spacing: 3) {
                        Text(hasAutomation
                            ? AutomationTriggerDisplay.summary(draft.trigger)
                            : "No automation")
                            .font(.subheadline)
                            .foregroundStyle(.white)
                            .lineLimit(1)
                        if hasAutomation, let bound = automationDevice {
                            Text(LaunchVocabulary.deviceCaption(bound))
                                .font(.caption)
                                .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                                .lineLimit(1)
                        }
                    }
                    Spacer(minLength: 0)
                }
            }
        } header: {
            GlassSectionHeader("Automation")
        } footer: {
            if hasAutomation {
                Text("The agent sets this up after writing the action.")
            }
        }
        .listRowBackground(glassFormRowFill)
    }

    /// The pushed detail: the same trigger rows and machine/agent options the
    /// automation form sheet renders, plus the on/off switch.
    private var automationDetail: some View {
        Form {
            Section {
                Toggle("Automate this action", isOn: $hasAutomation)
            } footer: {
                if automationDevices.isEmpty {
                    Text(AutomationCopy.noAutomationDevice)
                }
            }
            .listRowBackground(glassFormRowFill)

            if hasAutomation {
                AutomationTriggerForm(draft: $draft, options: filterOptions)
                LaunchOptionsSection(
                    variant: .automation,
                    devices: automationDevices,
                    deviceId: automationDeviceBinding,
                    noDeviceNote: AutomationCopy.noAutomationDevice,
                    availableAgents: LaunchVocabulary.agents(of: automationDevice),
                    agent: automationAgent,
                    onAgentChange: selectAutomationAgent,
                    model: $automationModel,
                    effort: $automationEffort
                )
            }
        }
        .scrollContentBackground(.hidden)
        .navigationTitle("Automation")
        .navigationBarTitleDisplayMode(.inline)
        .listSectionSpacing(8)
        .tint(DesignTokens.Palette.primary)
        // The machine pool can land after the sheet did — seed on the way in.
        .onAppear { seedAutomationAgent() }
    }

    // MARK: - Devices

    private var noDeviceNote: String {
        "No desktop online. Open the Exponential desktop app to start a run."
    }

    private var deviceBinding: Binding<String> {
        Binding(
            get: { device?.deviceId ?? "" },
            set: { value in
                let switched = value != lastSeededDeviceId
                deviceId = value
                clampAgentToDevice()
                if switched { applyDeviceDefaults() }
            }
        )
    }

    /// The machine that will RUN the automation, resolved off its own pool.
    private var automationDevice: SteerDevice? {
        if !automationDeviceId.isEmpty,
           let match = automationDevices.first(where: { $0.deviceId == automationDeviceId }) {
            return match
        }
        // EXP-622: the caller's default machine, when it is still a candidate.
        return automationDevices.first(where: \.isDefaultDevice) ?? automationDevices.first
    }

    /// Switching machines re-seeds the agent (EXP-615) rather than dropping the
    /// pin — a binding always names a concrete agent now.
    private var automationDeviceBinding: Binding<String> {
        Binding(
            get: { automationDevice?.deviceId ?? "" },
            set: { value in
                guard value != automationDevice?.deviceId else { return }
                automationDeviceId = value
                seedAutomationAgent()
            }
        )
    }

    /// Seed (and re-seed) the automation's agent off its bound machine: an
    /// unset pin, or one the newly picked machine cannot run, falls back to
    /// that machine's own default launch agent. A pin it CAN run is left
    /// alone, so a manual pick sticks.
    private func seedAutomationAgent() {
        guard let bound = automationDevice else { return }
        let agents = LaunchVocabulary.agents(of: bound)
        guard automationAgent.isEmpty || !agents.contains(automationAgent) else { return }
        automationAgent = LaunchVocabulary.defaultAgent(of: bound)
        automationModel = LaunchVocabulary.cliDefault
        automationEffort = LaunchVocabulary.cliDefault
    }

    private func selectAutomationAgent(_ value: String) {
        guard value != automationAgent else { return }
        automationAgent = value
        // The model/effort vocabularies are per agent — a switch has to reset
        // them, or a stale value hits a server refusal.
        automationModel = LaunchVocabulary.cliDefault
        automationEffort = LaunchVocabulary.cliDefault
    }

    // MARK: - Launch options (EXP-437, the Start-coding sheet's rules)

    private func selectAgent(_ value: String) {
        guard value != agent else { return }
        agent = value
        applyAgentDefaults(for: value)
    }

    private func clampAgentToDevice() {
        if !availableAgents.contains(agent) {
            selectAgent(availableAgents.first ?? "claude")
        }
    }

    private func seed() {
        guard !seeded else { return }
        seeded = true
        descriptionText = prefillDescription
        icon = prefillIcon
        if let prefillAutomation {
            hasAutomation = true
            draft = AutomationDraft(trigger: prefillAutomation)
        }
        applyDeviceDefaults()
        seedAutomationAgent()
    }

    private func applyDeviceDefaults() {
        if let advertised = device?.defaultLaunchAgent {
            agent = advertised
        } else if !availableAgents.contains(agent) {
            agent = availableAgents.first ?? "claude"
        }
        applyAgentDefaults(for: agent)
        lastSeededDeviceId = device?.deviceId
    }

    private func applyAgentDefaults(for value: String) {
        let advertised = device?.agentDefaults(for: value)
        model = seedModel(advertised?.model, for: value)
        effort = seedEffort(advertised?.effort, for: value)
        ultracode = advertised?.ultracode ?? false
        planMode = advertised?.planMode ?? false
        if agent != "claude" { ultracode = false }
        if !LaunchVocabulary.supportsPlanMode(agent) { planMode = false }
    }

    private func seedModel(_ value: String?, for agent: String) -> String {
        guard let value, !value.isEmpty,
              LaunchVocabulary.modelValues(for: agent).contains(value)
        else {
            return LaunchVocabulary.defaultModel(for: agent)
        }
        return value
    }

    private func seedEffort(_ value: String?, for agent: String) -> String {
        guard let value, LaunchVocabulary.effortValues(for: agent).contains(value) else {
            return LaunchVocabulary.cliDefault
        }
        return value
    }

    // MARK: - Load / submit

    @MainActor
    private func load() async {
        repos = (try? await deps.repositoriesApi.list(accountId: accountId, teamId: teamId)) ?? []
        filterOptions = await AutomationFilterOptions.load(
            db: deps.db, accountId: accountId, teamId: teamId
        )
    }

    /// The automation the creator agent should set up, or nil when the row is
    /// off or no machine can run one.
    private var configuredAutomation: AutomationSpec? {
        guard hasAutomation, let bound = automationDevice else { return nil }
        // A blank model/effort is the "CLI default" that stores NULL.
        let pinned = automationAgent.isEmpty ? nil : automationAgent
        let blankModel = automationModel.isEmpty || automationModel == LaunchVocabulary.cliDefault
        let blankEffort = automationEffort.isEmpty || automationEffort == LaunchVocabulary.cliDefault
        return AutomationSpec(
            trigger: draft.trigger,
            deviceId: bound.deviceId,
            agent: pinned,
            model: pinned == nil || blankModel ? nil : automationModel,
            effort: pinned == nil || blankEffort ? nil : automationEffort
        )
    }

    private func submit() {
        guard let device, canSubmit else { return }
        let action = builtin
        let options = SteerStartOptions(
            agent: agent,
            model: model == LaunchVocabulary.cliDefault ? "" : model,
            effort: effort == LaunchVocabulary.cliDefault ? "" : effort,
            ultracode: agent == "claude" ? ultracode : nil,
            planMode: LaunchVocabulary.supportsPlanMode(agent) ? planMode : nil
        )
        // Values in wire form: text trimmed, blank optionals dropped — the
        // very same helper the action inputs go through.
        var values = ActionInputValues.wireValues(
            action.inputs ?? [],
            values: [
                "description": descriptionText,
                "name": name,
                "repo": repoId,
                "icon": icon,
            ]
        )
        // EXP-583: a configured automation rides the description as a
        // machine-readable trailing block the creator agent copies verbatim
        // into `exponential_automations_create` (byte-identical across the
        // four clients — see AutomationNote.format).
        if let spec = configuredAutomation {
            values["description"] = (values["description"] ?? "") + AutomationNote.format(spec)
        }
        dismiss()
        onSubmit(device, action, options, values)
    }
}

File: apps/android/app/src/main/java/com/exponential/app/ui/actions/CreateActionSheet.kt
package com.exponential.app.ui.actions

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.exponential.app.data.api.ActionDto
import com.exponential.app.data.api.SteerDevice
import com.exponential.app.data.api.SteerStartOptions
import com.exponential.app.data.api.builtinCreateAction
import com.exponential.app.domain.AutomationTrigger
import com.exponential.app.domain.DomainContract
import com.exponential.app.domain.formatAutomationBlock
import com.exponential.app.domain.triggerSummary
import com.exponential.app.ui.components.DEFAULT_AGENT
import com.exponential.app.ui.components.GlassSheet
import com.exponential.app.ui.components.GlassTextField
import com.exponential.app.ui.components.GroupDivider
import com.exponential.app.ui.components.IconPicker
import com.exponential.app.ui.components.LaunchOptionsSection
import com.exponential.app.ui.components.LaunchOptionsVariant
import com.exponential.app.ui.components.OptionGroup
import com.exponential.app.ui.components.PickerRow
import com.exponential.app.ui.components.SheetHeight
import com.exponential.app.ui.components.SheetPrimaryAction
import com.exponential.app.ui.components.agentSeed
import com.exponential.app.ui.components.availableAgentsFor
import com.exponential.app.ui.components.defaultAgentFor
import com.exponential.app.ui.components.deviceOptionLabel
import com.exponential.app.ui.components.supportsPlanMode
import com.exponential.app.ui.icons.ExpIcons
import com.exponential.app.ui.issue.StartCodingSheetViewModel
import com.exponential.app.ui.theme.TextEmphasis

// The "New action" sheet (EXP-615, split out of the Start-coding sheet's old
// create mode): ONE form for authoring an action — icon + name, what it should
// do, the repository it clones, an always-visible Automation row, and the same
// device/agent/model/effort block every launch surface renders. Submitting
// RUNS the "Create action" builtin on the chosen machine: the agent writes the
// action for the team (and, when an automation is configured, copies the
// trailing note verbatim into `exponential_automations_create`).

@Composable
fun CreateActionSheet(
    teamId: String,
    devices: List<SteerDevice>,
    onRunAction: (SteerDevice, ActionDto, SteerStartOptions, Map<String, String>) -> Unit,
    onDismiss: () -> Unit,
    // EXP-530: a used suggestion's description + icon, keyed by input def key.
    prefilledInputs: Map<String, String>? = null,
    // EXP-583: the trigger an "Action + automation" suggestion proposes — it
    // seeds the Automation row, which is editable and clearable like any other.
    suggestionAutomation: AutomationTrigger? = null,
    dataViewModel: StartCodingSheetViewModel = hiltViewModel(),
) {
    val teamRepos by dataViewModel.repos.collectAsStateWithLifecycle()
    val boardOptions by dataViewModel.boardOptions.collectAsStateWithLifecycle()
    val labelOptions by dataViewModel.labelOptions.collectAsStateWithLifecycle()
    val statusOptions by dataViewModel.statusOptions.collectAsStateWithLifecycle()
    val automationDevices by dataViewModel.automationDevices.collectAsStateWithLifecycle()

    val createAction = remember(teamId) { builtinCreateAction(teamId) }

    // ── The form ─────────────────────────────────────────────────────────────
    var name by remember { mutableStateOf(prefilledInputs?.get("name").orEmpty()) }
    var description by remember { mutableStateOf(prefilledInputs?.get("description").orEmpty()) }
    var repoId by remember { mutableStateOf("") }
    var icon by remember { mutableStateOf(prefilledInputs?.get("icon").orEmpty()) }

    // ── The automation (EXP-583) ─────────────────────────────────────────────
    // Always offered now: the row reads "No automation" until one is set, and
    // the detail sheet's leading "None" segment clears it again. The machine
    // and the agent pins are the AUTOMATION's, independent of the desktop
    // running the creator run.
    var automation by remember {
        mutableStateOf(
            suggestionAutomation?.let { automationDraftFor(it) }
                ?: AutomationDraft(kind = AUTOMATION_KIND_NONE),
        )
    }
    var automationOpen by remember { mutableStateOf(false) }

    // Pre-pick the first automation-capable machine once the rows sync; a
    // manual re-pick sticks (the draft already carries a deviceId then).
    LaunchedEffect(automationDevices) {
        if (automation.deviceId != null) return@LaunchedEffect
        // EXP-622: the caller's default machine, else the first candidate.
        (automationDevices.firstOrNull { it.isDefault } ?: automationDevices.firstOrNull())
            ?.let { automation = automation.copy(deviceId = it.deviceId) }
    }

    // ── The creator run's machine + options (EXP-437 seeding) ────────────────
    // EXP-672: online with a runnable agent is the whole rule — every build
    // above the version floor runs the builtin.
    val candidates = remember(devices) {
        devices.filter { it.online && it.hasRunnableAgent }
    }
    val initialDevice = remember {
        // EXP-622: the caller's default machine, else the first candidate.
        candidates.firstOrNull { it.isDefault } ?: candidates.firstOrNull()
    }
    val initialAgent = remember { defaultAgentFor(initialDevice) }
    val initialSeed = remember { agentSeed(initialDevice, initialAgent) }

    var deviceId by remember { mutableStateOf(initialDevice?.deviceId) }
    var agent by remember { mutableStateOf(initialAgent) }
    var model by remember { mutableStateOf(initialSeed.model) }
    var effort by remember { mutableStateOf(initialSeed.effort) }
    var ultracode by remember { mutableStateOf(initialSeed.ultracode) }
    var planMode by remember { mutableStateOf(initialSeed.planMode) }

    val device = candidates.firstOrNull { it.deviceId == deviceId }
        ?: candidates.firstOrNull { it.isDefault }
        ?: candidates.firstOrNull()
    val availableAgents = availableAgentsFor(device)

    fun applyAgentSeed(next: String) {
        agent = next
        val seed = agentSeed(device, next)
        model = seed.model
        effort = seed.effort
        ultracode = seed.ultracode
        planMode = seed.planMode
    }

    fun selectAgent(next: String) {
        if (next == agent) return
        applyAgentSeed(next)
    }

    // A machine settling late (or a stricter candidate filter) re-seeds the
    // options once; the latch keeps a re-emit from stomping manual edits.
    var seededDeviceId by remember { mutableStateOf(initialDevice?.deviceId) }
    LaunchedEffect(device?.deviceId) {
        val settled = device ?: return@LaunchedEffect
        if (seededDeviceId == settled.deviceId) {
            if (agent !in availableAgentsFor(settled)) {
                selectAgent(availableAgentsFor(settled).firstOrNull() ?: DEFAULT_AGENT)
            }
            return@LaunchedEffect
        }
        seededDeviceId = settled.deviceId
        applyAgentSeed(defaultAgentFor(settled))
    }

    val canCreate = device != null && description.isNotBlank()

    GlassSheet(
        title = "New action",
        onDismiss = onDismiss,
        modifier = Modifier.testTag("create-action-sheet"),
        height = SheetHeight.Full,
        primaryAction = SheetPrimaryAction(
            label = "Create",
            enabled = canCreate,
            icon = ExpIcons.actionCreate,
            onClick = create@{
                val target = device ?: return@create
                val options = SteerStartOptions(
                    model = model,
                    effort = effort,
                    ultracode = if (agent == DEFAULT_AGENT) ultracode else null,
                    planMode = if (supportsPlanMode(agent)) planMode else null,
                    agent = agent,
                )
                // EXP-583: the configured automation rides the description as
                // a machine-readable note the creator agent copies verbatim
                // into exponential_automations_create.
                val automationDeviceId = automation.deviceId
                val trigger = automationDraftToTrigger(automation)
                val body = if (trigger != null && automationDeviceId != null) {
                    description + formatAutomationBlock(
                        trigger,
                        deviceId = automationDeviceId,
                        agent = automation.agent,
                        model = automation.model,
                        effort = automation.effort,
                    )
                } else {
                    description
                }
                val inputs = buildMap {
                    put("description", body)
                    if (name.isNotBlank()) put("name", name.trim())
                    if (repoId.isNotEmpty()) put("repo", repoId)
                    if (icon.isNotEmpty()) put("icon", icon)
                }
                onRunAction(target, createAction, options, inputs)
                onDismiss()
            },
        ),
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .weight(1f)
                .verticalScroll(rememberScrollState()),
        ) {
            // Icon + name on one row (web/desktop/iOS parity), then the
            // description that the creator agent actually works from.
            OptionGroup {
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 16.dp, vertical = 8.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    IconPicker(selected = icon, onSelect = { icon = it }, allowsNone = true)
                    Spacer(Modifier.width(12.dp))
                    GlassTextField(
                        value = name,
                        onValueChange = {
                            name = it.take(DomainContract.actionInputTextMax)
                        },
                        modifier = Modifier.weight(1f),
                        placeholder = "Name (optional)",
                        singleLine = true,
                    )
                }
            }
            Spacer(Modifier.height(8.dp))

            // EXP-694 (S7): the brief sits INSIDE the card stack like the edit
            // sheet's description and prompt — the field's own fill/hairline
            // would double the group's chrome, so it drops both and the
            // placeholder carries the title.
            OptionGroup {
                GlassTextField(
                    value = description,
                    onValueChange = {
                        description = it.take(DomainContract.actionInputTextMax)
                    },
                    modifier = Modifier.fillMaxWidth(),
                    placeholder = "What should this action do?",
                    minLines = 4,
                    bordered = false,
                )
            }
            Spacer(Modifier.height(8.dp))

            OptionGroup {
                PickerRow(
                    label = "Repository",
                    value = when {
                        repoId.isEmpty() -> "None"
                        else -> teamRepos.firstOrNull { it.id == repoId }?.fullName ?: repoId
                    },
                    options = listOf("") + teamRepos.map { it.id },
                    selected = repoId,
                    optionLabel = { id ->
                        if (id.isEmpty()) {
                            "None"
                        } else {
                            teamRepos.firstOrNull { it.id == id }?.fullName ?: id
                        }
                    },
                    onSelect = { repoId = it },
                )
                GroupDivider()
                // The automation summary row — always here, "No automation"
                // until one is configured (web parity).
                AutomationSummaryRow(
                    draft = automation,
                    devices = automationDevices,
                    onClick = { automationOpen = true },
                )
            }
            Spacer(Modifier.height(8.dp))

            LaunchOptionsSection(
                variant = LaunchOptionsVariant.Launch,
                devices = candidates,
                device = device,
                onDeviceChange = { id ->
                    deviceId = id
                    val candidate = candidates.firstOrNull { it.deviceId == id }
                    val available = availableAgentsFor(candidate)
                    if (agent !in available) {
                        selectAgent(available.firstOrNull() ?: DEFAULT_AGENT)
                    }
                },
                agent = agent,
                availableAgents = availableAgents,
                onAgentChange = ::selectAgent,
                model = model,
                onModelChange = { model = it },
                effort = effort,
                onEffortChange = { effort = it },
                noDeviceNote = "No desktop online. Open the Exponential desktop app " +
                    "to start a run.",
                ultracode = ultracode,
                onUltracodeChange = { ultracode = it },
                planMode = planMode,
                onPlanModeChange = { planMode = it },
            )
            Spacer(Modifier.height(24.dp))
        }
    }

    // The automation detail, nested in its own glass sheet (EXP-607 pattern) —
    // the trigger's leading "None" segment IS the clear affordance.
    if (automationOpen) {
        GlassSheet(
            title = "Automation",
            onDismiss = { automationOpen = false },
            height = SheetHeight.Full,
            primaryAction = SheetPrimaryAction(label = "Done", onClick = { automationOpen = false }),
        ) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .verticalScroll(rememberScrollState()),
            ) {
                AutomationTriggerFields(
                    draft = automation,
                    boards = boardOptions,
                    labels = labelOptions,
                    statuses = statusOptions,
                    onChange = { automation = it },
                    allowNone = true,
                )
                if (automation.kind != AUTOMATION_KIND_NONE) {
                    Spacer(Modifier.height(8.dp))
                    AutomationBindingFields(
                        draft = automation,
                        devices = automationDevices,
                        onChange = { automation = it },
                    )
                }
                Spacer(Modifier.height(8.dp))
            }
        }
    }
}

/**
 * The always-visible automation row: the configured trigger's sentence plus
 * the machine it would run on, or "No automation" when there is none.
 */
@Composable
private fun AutomationSummaryRow(
    draft: AutomationDraft,
    devices: List<SteerDevice>,
    onClick: () -> Unit,
) {
    val trigger = automationDraftToTrigger(draft)
    val boundDevice = devices.firstOrNull { it.deviceId == draft.deviceId }
    val summary = when {
        trigger == null -> "No automation"
        boundDevice == null -> triggerSummary(trigger)
        else -> "${triggerSummary(trigger)} · ${deviceOptionLabel(boundDevice)}"
    }
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            "Automation",
            style = MaterialTheme.typography.bodyLarge,
            color = MaterialTheme.colorScheme.onSurface,
            modifier = Modifier.weight(1f),
        )
        Text(
            summary,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Secondary),
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            modifier = Modifier.weight(1f, fill = false),
        )
        Icon(
            ExpIcons.uiChevronRight,
            contentDescription = null,
            modifier = Modifier.size(20.dp),
            tint = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
        )
    }
}
